File: data/src/commonMain/kotlin/com/example/myapplication/datasource/ArenaTournamentDatasource.kt
package com.example.myapplication.datasource

import com.example.myapplication.rawresponses.*
import com.example.myapplication.rawresponses.createresponses.*
import io.ktor.http.Url

interface ArenaTournamentDatasource {

    suspend fun createGameMode(gameModeJSON: CreateGameModeJSON): ModeJSON
    suspend fun createGame(gameJSON: CreateGameJSON): GameJSON
    suspend fun createRegistration(registrationJSON: CreateRegistrationJSON): RegistrationJSON
    suspend fun createTournament(tournamentJSON: CreateTournamentJSON): TournamentJSON
    suspend fun createUser(userJSON: CreateUserJSON): UserJSON

    suspend fun getAllGames(page: Int): MultipleGamesJSON
    suspend fun getGameByName(gameName: String): GameJSON
    suspend fun searchGamesByName(query: String, page: Int): MultipleGamesJSON
    suspend fun getGameByLink(link: String): GameJSON
    suspend fun getGamesContainingName(gameName: String, page: Int): MultipleGamesJSON
    suspend fun getGamesByMode(mode: String, page: Int): MultipleGamesJSON

    suspend fun getAllTournaments(page: Int): MultipleTournamentsJSON
    suspend fun getTournamentById(id: Long): TournamentJSON
    suspend fun getTournamentByLink(link: String): TournamentJSON
    suspend fun getTournamentsByMode(mode: String, page: Int): MultipleTournamentsJSON
    suspend fun getTournamentsByGameName(gameName: String, page: Int): MultipleTournamentsJSON
    suspend fun getTournamentsByUser(userId: String, page: Int): MultipleTournamentsJSON
    suspend fun searchTournamentsByName(name: String, page: Int): MultipleTournamentsJSON
    suspend fun getShowCaseTournaments(page: Int): MultipleTournamentsJSON
    suspend fun getTournamentsContainingTitle(title: String, page: Int): MultipleTournamentsJSON

    suspend fun getRegistrationById(id: Long): RegistrationJSON
    suspend fun getRegistrationByLink(link: String): RegistrationJSON
    suspend fun getRegistrationsByUser(userId: String, page: Int): MultipleRegistrationsJSON
    suspend fun getRegistrationsByTournament(tournamentId: Long, page: Int): MultipleRegistrationsJSON

    suspend fun getUserById(id: String): UserJSON
    suspend fun getUserByLink(link: String): UserJSON
    suspend fun getAccountVerificationStatus(): AccountStatusJSON
    suspend fun getAccountSubscription(): SubscriptionStatusJSON
    suspend fun searchTournaments(title: String, gameId: String?, page: Int): MultipleTournamentsJSON

    interface Endpoints {

        val protocol: String
        val host: String
        val port: Int

        fun createGameModeUrl(): Url
        fun createGameUrl(): Url
        fun createRegistrationUrl(): Url
        fun createTournamentUrl(): Url
        fun createUserUrl(): Url


        fun allGamesUrl(page: Int): Url
        fun gameByNameUrl(name: String): Url
        fun searchGamesByNameUrl(query: String, page: Int): Url
        fun gamesContainingNameUrl(gameName: String, page: Int): Url
        fun gamesByModeUrl(mode: String, page: Int): Url


        fun allTournamentsUrl(page: Int): Url
        fun tournamentByIdUrl(id: Long): Url
        fun tournamentsByGameName(gameName: String, page: Int): Url
        fun tournamentsByModeUrl(mode: String, page: Int): Url
        fun tournamentsByAdmin(userId: String, page: Int): Url
        fun searchTournamentsByNameUrl(query: String, page: Int): Url
        fun getShowCaseTournaments(page: Int): Url
        fun getTournamentsContainingTitle(title: String, page: Int): Url

        fun allRegistrationsUrl(page: Int): Url
        fun registrationByIdUrl(id: Long): Url
        fun registrationsByUserUrl(userId: String, page: Int): Url
        fun registrationsByTournamentUrl(tournamentId: Long, page: Int): Url

        fun userByIdUrl(userId: String): Url
        fun isAccountVerifiedUrl(): Url
        fun isAccountSubscribedUrl(): Url
        fun searchTournaments(title: String, gameId: String?, page: Int): Url

    }
}

File: web-client/src/app/data/datasources/arena-tournament-datasource.ts
import {CreateGameModeJSON} from '../rawresponses/creation/create-game-mode-json';
import {ModeJSON} from '../rawresponses/single/mode-json';
import {CreateGameJSON} from '../rawresponses/creation/create-game-json';
import {GameJSON} from '../rawresponses/single/game-json';
import {CreateRegistrationJSON} from '../rawresponses/creation/create-registration-json';
import {RegistrationJSON} from '../rawresponses/single/registration-json';
import {CreateTournamentJSON} from '../rawresponses/creation/create-tournament-json';
import {TournamentJSON} from '../rawresponses/single/tournament-json';
import {MultipleGamesJSON} from '../rawresponses/multiple/multiple-games-json';
import {MultipleTournamentsJSON} from '../rawresponses/multiple/multiple-tournaments-json';
import {MultipleRegistrationsJSON} from '../rawresponses/multiple/multiple-registrations-json';
import {UserJSON} from '../rawresponses/single/user-json';
import {AccountStatusJSON} from '../rawresponses/single/account-status-json';
import {SubscriptionStatusJSON} from '../rawresponses/single/subscription-status-json';
import {Observable} from 'rxjs';

export abstract class ArenaTournamentDatasource {

  abstract createGameMode(createGameModeJSON: CreateGameModeJSON): Observable<ModeJSON>;

  abstract createGame(createGameJSON: CreateGameJSON): Observable<GameJSON>;

  abstract createRegistration(createRegistrationJSON: CreateRegistrationJSON): Observable<RegistrationJSON>;

  abstract createTournament(createTournamentJSON: CreateTournamentJSON): Observable<TournamentJSON>;

  abstract getAllGames(page: number): Observable<MultipleGamesJSON>;

  abstract getGameByName(gameName: string): Observable<GameJSON>;

  abstract searchGamesByName(query: string, page: number): Observable<MultipleGamesJSON>;

  abstract getGameByLink(link: string): Observable<GameJSON>;

  abstract getGamesContainingName(gameName: string, page: number): Observable<MultipleGamesJSON>;

  abstract getGamesByMode(mode: string, page: number): Observable<MultipleGamesJSON>;

  abstract getAllTournaments(page: number): Observable<MultipleTournamentsJSON>;

  abstract getTournamentById(id: number): Observable<TournamentJSON>;

  abstract getTournamentByLink(link: string): Observable<TournamentJSON>;

  abstract getTournamentsByMode(mode: string, page: number): Observable<MultipleTournamentsJSON>;

  abstract getTournamentsByGameName(gameName: string, page: number): Observable<MultipleTournamentsJSON>;

  abstract getTournamentsByUser(userId: string, page: number): Observable<MultipleTournamentsJSON>;

  abstract searchTournamentsByName(name: string, page: number): Observable<MultipleTournamentsJSON>;

  abstract getShowCaseTournaments(page: number): Observable<MultipleTournamentsJSON>;

  abstract getTournamentsContainingTitle(title: string, page: number): Observable<MultipleTournamentsJSON>;

  abstract getRegistrationById(id: number): Observable<RegistrationJSON>;

  abstract getRegistrationByLink(link: string): Observable<RegistrationJSON>;

  abstract getRegistrationsByUser(userId: string, page: number): Observable<MultipleRegistrationsJSON>;

  abstract getRegistrationsByTournament(tournamentId: number, page: number): Observable<MultipleRegistrationsJSON>;

  abstract getUserById(id: string): Observable<UserJSON>;

  abstract getUserByLink(link: string): Observable<UserJSON>;

  abstract getAccountVerificationStatus(): Observable<AccountStatusJSON>;

  abstract getAccountSubscription(): Observable<SubscriptionStatusJSON>;

  abstract searchTournaments(title: string, page: number): Observable<MultipleTournamentsJSON>;
}
